yara-x-cli/src/commands/dump.rs

```diff
@@ -0,0 +1,186 @@
+use clap::{
+    arg, value_parser, Arg, ArgAction, ArgMatches, Command, ValueEnum,
+};
+
+use colored_json::ToColoredJson;
+use protobuf::{reflect::ReflectValueRef::Bool, MessageDyn};
+use protobuf_json_mapping::print_to_string;
+use std::fmt::Write;
+use std::fs::File;
+use std::io::stdin;
+use std::io::Read;
+use std::path::PathBuf;
+use yansi::{Color::Cyan, Paint};
+use yara_x_proto::exts::module_options;
+
+use yara_x_dump::Dumper;
+
+use yara_x::get_builtin_modules_names;
+
+#[derive(Debug, Clone, ValueEnum)]
+enum OutputFormats {
+    Json,
+    Yaml,
+}
+
+/// Creates the `dump` command.
+/// The `dump` command dumps information about binary files.
+///
+/// # Returns
+///
+/// Returns a `Command` struct that represents the `dump` command.
+pub fn dump() -> Command {
+    super::command("dump")
+        .about("Dump information about binary files")
+        .arg(
+            arg!(<FILE>)
+                .help("Path to binary file")
+                .value_parser(value_parser!(PathBuf))
+                .required(false),
+        )
+        .arg(
+            arg!(-o --"output-format" <FORMAT>)
+                .help("Desired output format")
+                .value_parser(value_parser!(OutputFormats))
+                .required(false),
+        )
+        .arg(
+            arg!(-c - -"color")
+                .help("Use colorful output")
+        )
+        .arg(
+            Arg::new("modules")
+            .long("modules")
+            .short('m')
+            .help("Name of the module or comma-separated list of modules to be used for parsing")
+            .required(false)
+            .action(ArgAction::Append)
+            .value_parser(get_builtin_modules_names()),
+        )
+}
+
+// Checks if the module output is valid by checking the validity flag.
+//
+// # Arguments
+//
+// * `mod_output`: The module output to check.
+//
+// # Returns
+//
+// * `true` if the module output is valid, `false` otherwise.
+fn module_is_valid(mod_output: &dyn MessageDyn) -> bool {
```

Comment:
I'm not quite convinced about all this logic that determines if the output of a module is valid. It requires that you mark a field in the proto and makes things more complex with too little gain. If you the user uses the command-line for dumping the result of the `pe` module and he passes a non-PE file, the result will be a very small YAML/JSON where the `is_pe` field is set to `false`, that's ok to me.

The "auto" option, which allows to pass only the file and let the CLI figure out which module make sense to be taken into account and which not, is a nice addition, but it can be implemented by putting the logic in the CLI tool itself, instead of having to bake all that validity check into `yara-x` itself.

The steps the CLI would take are:

1. Given some `&[u8]` with the content of the file, ask `yara-x` to produce the output for module "pe" (or any other module name). This is the logic I'm implementing now, and will return a protobuf. 
2. As the CLI already have access to the protobuf, it the validation logic can be executed by the CLI before passing the protobuf to the serializer of obtaining a text representation of it.
3. Based in the output format requested, use `protobuf_json_mapping::print_to_string` or our own YAML serializer.

Reply:
> If you the user uses the command-line for dumping the result of the pe module and he passes a non-PE file, the result will be a very small YAML/JSON where the is_pe field is set to false, that's ok to me.

This is already implemented, if the user manually specifies the module he wants to use. The output will be from given module even when `is_pe` or any other flag is set to false or not set at all. All complexity comes from automatic selection. Some modules as `pe` or `lnk` has a validity flag built-in (`is_pe` or `is_lnk`) but macho (or others don't have this option. Therefore it is right now a bit module specific and I wanted some universal solution by marking certain field as a flag. This is unified across all modules and easy to validate. But I am open to another solutions that would not require to add another specification to protobuf itself.

>As the CLI already have access to the protobuf, it the validation logic can be executed by the CLI before passing the protobuf to the serializer of obtaining a text representation of it.

I am not sure what is the dependancy to `yara-x` right now. The `module_is_valid()` function takes protobuf and just checks the flag. I am not sure what would be another option. Without this flag I can manually traverse the protobuf and search for some kind of field that would tell me "ok this is valid" but as it differs across the modules it would be a bit complex and with a new module there is a chance that something would have to be added into the code.

Reply:
> This is already implemented, if the user manually specifies the module he wants to use. The output will be from given module even when is_pe or any other flag is set to false or not set at all. All complexity comes from automatic selection. Some modules as pe or lnk has a validity flag built-in (is_pe or is_lnk) but macho (or others don't have this option. Therefore it is right now a bit module specific and I wanted some universal solution by marking certain field as a flag. This is unified across all modules and easy to validate. But I am open to another solutions that would not require to add another specification to protobuf itself.

I like the feature. I like that the `dump` command produces only the output for the file types that make sense, instead of blindly dumping the results of all modules. What I don't like is that this functionality is embedded in `yara-x`, by means of the having to declare in the `.proto` a field that determines whether the output is valid or not.
I believe that all those checks also correspond to the CLI tool itself, not to the `yara-x` library.

> I am not sure what is the dependancy to yara-x right now. The module_is_valid() function takes protobuf and just checks the flag. I am not sure what would be another option. Without this flag I can manually traverse the protobuf and search for some kind of field that would tell me "ok this is valid" but as it differs across the modules it would be a bit complex and with a new module there is a chance that something would have to be added into the code.

What I mean with dependency to `yara-x` is that the CLI tool, in order to determine which module it should include in the output, is relying on the validity flag provided by the `yara-x` library itself. This validity logic is something that probably only the CLI will use, and its better implemented at the CLI level. The problem you have now is that it's imposible to inspect the result produced by each at the CLI level, and therefore you don't have access to the `is_pe` or `is_lnk` at the CLI level, but that's one of the problems I plan to solve in the API I'm working on.